test/interface.jl

```diff
@@ -0,0 +1,22 @@
+
+using SymbolicUtils: Term, to_symbolic
+SymbolicUtils.symtype(::Union{Expr, Symbol}) = Number
+SymbolicUtils.istree(ex::Expr) = ex.head == :call
+SymbolicUtils.operation(ex::Expr) = ex.args[1]
+SymbolicUtils.arguments(ex::Expr) = ex.args[2:end]
+
+for f ∈ [:+, :-, :*, :/, :^]
+    @eval begin
+        Base.$f(x::Union{Expr, Symbol}, y::Number) = Expr(:call, $f, x, y) |> to_symbolic
+        Base.$f(x::Number, y::Union{Expr, Symbol}) = Expr(:call, $f, x, y) |> to_symbolic
+        Base.$f(x::Union{Expr, Symbol}, y::Union{Expr, Symbol}) = (Expr(:call, $f, x, y)) |> to_symbolic
```

Comment:
The main problem is that we need the `|>  to_symbolic` here.  If I get rid of it and try to do a simplification, this happens:
```julia
julia> ex = 1 + (:x - 2)
:((+)(1, (-)(x, 2)))

julia> SymbolicUtils.to_symbolic(ex)
((+)(x, -2) + 1)

julia> simplify(ex) |> dump
Term{Number}
  f: + (function of type typeof(+))
  arguments: Array{Any}((2,))
    1: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: + (function of type typeof(+))
        2: Symbol x
        3: Int64 -2
    2: Int64 1
```
Note that on the outside, this is a `Term`, but it's first argument is an `expr`
this is because the `@rule(-(~x, ~y) => ~x + -1(~y))` causes the operation `:x + -2` to happen which results in an `Expr`. I guess I can make it so that a `RuleSet` calls `to_symbolic` before every rule application instead of just once at the very start of `simplify`.

Reply:
Actually wait, we could just make the `matcher` system rely on `istree`, `operation` and `arguments` an then we might not even need to convert things to `Term`, right?

Reply:
Hmm matcher can definitely work with those 3 functions! But the problem is you need to construct `tree`s when rewriting, so we do that with the `Term` constructor. I didn't want to complicate that a whole lot.

https://github.com/SciML/ModelingToolkit.jl/pull/326/files#diff-fd73d8caed68246d177d495cbfff183dR25

this was required. Maybe we should just document `to_symbolic` here. I think we should document `to_symbolic`.

So for conversion you either need to define 1) `to_symbolic` or 2) `istree`, `arg`, `op` combo. For expression trees 2) is preferred because it will automatically infer types of the terms.

Reply:
I might be misunderstanding you here, but defining `to_symbolic` is not enough, because when you apply rules you get objects which are no longer `Term`s.

Reply:
Yeah `to_symbolic` is the function which converts any other IR into that of this package.
So if you see my link there is a `to_symbolic` method to convert to `SymbolicUtils.Variable`, you may need to define `to_symbolic(::Symbol)` here. `to_symbolic` should be the main part of the interface.

Reply:
Oh I see what you're saying now. I didn't see your comment until now, but I got everything working by just making `Rule`s apply `to_symbolic` before and after they're called which does work.